File: 01-january/16-max-area-by-removing-fences/solution.java
import java.util.*;

class Solution {
    public int maximizeSquareArea(int m, int n, int[] hFences, int[] vFences) {
        final long MOD = 1_000_000_007L;
        Set<Integer> hset = buildDiffSet(hFences, m);
        Set<Integer> vset = buildDiffSet(vFences, n);

        int best = -1;
        for (int d : hset) {
            if (vset.contains(d) && d > best) best = d;
        }
        if (best == -1) return -1;
        long area = (long) best * best;
        return (int) (area % MOD);
    }

    private Set<Integer> buildDiffSet(int[] arr, int limit) {
        int[] all = Arrays.copyOf(arr, arr.length + 2);
        all[arr.length] = 1;
        all[arr.length + 1] = limit;
        Arrays.sort(all);

        Set<Integer> res = new HashSet<>();
        for (int i = 0; i < all.length; i++) {
            for (int j = i + 1; j < all.length; j++) {
                res.add(all[j] - all[i]);
            }
        }
        return res;
    }
}

File: 01-january/16-max-area-by-removing-fences/solution.py
from typing import List

class Solution:
    def maximizeSquareArea(self, m: int, n: int, hFences: List[int], vFences: List[int]) -> int:
        MOD = 10**9 + 7

        def diff_set(arr: List[int], limit: int) -> set[int]:
            all_vals = arr[:] + [1, limit]
            all_vals.sort()
            res = set()
            for i in range(len(all_vals)):
                for j in range(i + 1, len(all_vals)):
                    res.add(all_vals[j] - all_vals[i])
            return res

        hset = diff_set(hFences, m)
        vset = diff_set(vFences, n)

        best = -1
        for d in hset:
            if d in vset and d > best:
                best = d

        if best == -1:
            return -1
        return (best * best) % MOD
